File: tools/lz77.py
"""
lz77.py — Descompresor LZSS de PS2 (verificado contra MIPS en SLPS_256.11).

Formato del header (12 bytes, NO 16):
  [magic "LZ77"(4)] [decomp_size:u32] [comp_size:u32]

El "metadata" NO es un campo separado del header. El decompressor del PS2
empieza a leer en el byte 12. Los primeros 4 bytes del stream comprimido
son lo que llamábamos "metadata". El campo comp_size indica el tamaño
total del stream desde el byte 12 (incluyendo esos 4 bytes iniciales).

Fix 2026-06-28: El decompressor nativo usa skip=12, no skip=16.
Esto es la causa raíz de la pantalla negra.

Algoritmo (confirmado por desensamblado del ELF en 0x3B500):
  - Ventana: 4096 bytes, inicio: 0xFEE, init: 0x00
  - Flags: LSB-first (bit 0 primero)
  - Bit=1 → LITERAL (1 byte)
  - Bit=0 → MATCH (2 bytes):
      offset = b1 | ((b2 & 0xF0) << 4)   (12 bits, absoluto en ventana)
      length = (b2 & 0x0F) + 3           (3-18 bytes)
"""

import struct

MAGIC = b"LZ77"
WINDOW_SIZE = 4096
WINDOW_START = 0xFEE
MIN_MATCH = 3
MAX_MATCH = 18


def decompress(compressed_data, expected_size=None, strict=True):
    data = compressed_data
    pos = 0

    if data[:4] == MAGIC:
        if len(data) < 12:
            raise ValueError("Header LZ77 truncado")
        expected_size = struct.unpack_from("<I", data, 4)[0]
        comp_size = struct.unpack_from("<I", data, 8)[0]
        stream_end = 12 + comp_size
        if len(data) < stream_end:
            if strict:
                raise ValueError(
                    f"Stream LZ77 truncado: hay {max(0, len(data) - 12):,} bytes, "
                    f"header comp_size={comp_size:,}"
                )
            stream_end = len(data)
        # Saltar header de 12 bytes (magic + decomp_size + comp_size) y limitar
        # estrictamente al comp_size real. Bytes posteriores son padding/otro archivo.
        data = data[12:stream_end]
    elif expected_size is None:
        raise ValueError("Se requiere expected_size o header LZ77 de 12 bytes")

    out = bytearray()
    # FIX 2026-06-29: El PS2 nativo inicializa la ventana con 0x00.
    # Confirmado por desensamblado MIPS: memset(buf, 0, 4113) en 0x0013B3D0
    # y por comparación directa contra RAM del savestate funcional (0 diffs).
    window = bytearray([0x00] * WINDOW_SIZE)
    window_pos = WINDOW_START

    while pos < len(data) and len(out) < expected_size:
        flags = data[pos]
        pos += 1

        for bit in range(8):
            if len(out) >= expected_size:
                break

            # LSB-first: bit 0 primero
            is_literal = (flags & (1 << bit)) != 0

            if is_literal:
                if pos >= len(data):
                    break
                val = data[pos]
                pos += 1
                out.append(val)
                window[window_pos] = val
                window_pos = (window_pos + 1) & 0xFFF
            else:
                if pos + 1 >= len(data):
                    break
                b1 = data[pos]
                b2 = data[pos + 1]
                pos += 2

                offset = b1 | ((b2 & 0xF0) << 4)
                length = (b2 & 0x0F) + 3

                for _ in range(length):
                    if len(out) >= expected_size:
                        break
                    val = window[offset]
                    out.append(val)
                    window[window_pos] = val
                    window_pos = (window_pos + 1) & 0xFFF
                    offset = (offset + 1) & 0xFFF

    if strict and len(out) != expected_size:
        raise ValueError(
            f"Descompresión incompleta: salida={len(out):,} bytes, "
            f"esperado={expected_size:,}"
        )

    return bytes(out)


def compress(uncompressed_data, all_literal=False):

    data_len = len(uncompressed_data)
    compressed = bytearray()

    window = bytearray([0x00] * WINDOW_SIZE)
    window_pos = WINDOW_START

    src_pos = 0
    block_flags = 0
    bit_count = 0
    block_tokens = bytearray()

    def flush_block():
        nonlocal block_flags, bit_count, block_tokens
        if bit_count > 0:
            compressed.append(block_flags)
            compressed.extend(block_tokens)
            block_flags = 0
            bit_count = 0
            block_tokens = bytearray()

    while src_pos < data_len:
        match_len = 0
        match_offset = 0
        max_len = min(MAX_MATCH, data_len - src_pos)

        if not all_literal and max_len >= MIN_MATCH:
            best_len = 0
            best_offset = 0

            # Búsqueda greedy de ventana completa (misma estrategia que el
            # compresor original del juego). Busca el match más largo en
            # toda la ventana deslizante de 4096 bytes.
            search_start = max(0, src_pos - WINDOW_SIZE)
            search_end = src_pos - MIN_MATCH
            search_range = min(search_end - search_start, WINDOW_SIZE)

            # Optimización: buscar desde la posición actual hacia atrás
            for dist in range(MIN_MATCH, min(search_range + 1, WINDOW_SIZE)):
                w_idx = (window_pos - dist) & 0xFFF
                search_max = min(max_len, dist)
                curr_len = 0
                while curr_len < search_max:
                    wpos = (w_idx + curr_len) & 0xFFF
                    if window[wpos] != uncompressed_data[src_pos + curr_len]:
                        break
                    curr_len += 1
                if curr_len > best_len:
                    best_len = curr_len
                    best_offset = w_idx
                    if best_len == max_len:
                        break

            match_len = best_len
            match_offset = best_offset

        if match_len < MIN_MATCH:
            match_len = 1
            val = uncompressed_data[src_pos]
            block_flags |= (1 << bit_count)  # bit=1 → literal
            block_tokens.append(val)
        else:
            b1 = match_offset & 0xFF
            b2 = ((match_offset >> 4) & 0xF0) | ((match_len - 3) & 0x0F)
            block_tokens.extend([b1, b2])

        for k in range(match_len):
            window[window_pos] = uncompressed_data[src_pos + k]
            window_pos = (window_pos + 1) & 0xFFF

        src_pos += match_len
        bit_count += 1
        if bit_count == 8:
            flush_block()

    flush_block()

    # Construir header de 12 bytes: [magic:4][decomp_size:4][comp_size:4]
    # El stream comprimido va a continuacion (bytes 12+)
    comp_stream = bytes(compressed)
    comp_size = len(comp_stream)
    header = struct.pack("<4sII", MAGIC, data_len, comp_size)
    return header + comp_stream


def decompress_file(input_path, output_path=None):
    from pathlib import Path
    data = Path(input_path).read_bytes()
    decomp = decompress(data)
    if output_path is None:
        output_path = str(input_path) + ".dec"
    Path(output_path).write_bytes(decomp)
    return decomp


def compress_file(input_path, output_path=None, all_literal=False):
    from pathlib import Path
    data = Path(input_path).read_bytes()
    comp = compress(data, all_literal=all_literal)
    if output_path is None:
        output_path = str(input_path) + ".lz77"
    Path(output_path).write_bytes(comp)
    return comp


if __name__ == "__main__":
    import sys
    from pathlib import Path

    if len(sys.argv) < 2:
        print("Uso: python lz77.py [d|decompress] <input> [output]")
        print("      python lz77.py [c|compress] <input> [output]  (solo para tests locales)")
        sys.exit(1)

    mode = sys.argv[1]
    input_path = sys.argv[2]
    output_path = sys.argv[3] if len(sys.argv) > 3 else None

    if mode in ("d", "decompress"):
        result = decompress_file(input_path, output_path)
        print(f"Descomprimido: {len(result):,} bytes")
    elif mode in ("c", "compress"):
        result = compress_file(input_path, output_path)
        print(f"Comprimido: {len(result):,} bytes  [ADVERTENCIA: no usar para ISO]")
    else:
        print(f"Modo desconocido: {mode}")

File: crates/straw-core/src/lz77.rs
use anyhow::{bail, Result};

const MAGIC: &[u8; 4] = b"LZ77";
const WINDOW_SIZE: usize = 4096;
const WINDOW_START: usize = 0xFEE;
const MIN_MATCH: usize = 3;
const MAX_MATCH: usize = 18;

pub fn decompress_lz77(data: &[u8], expected_size: Option<usize>, strict: bool) -> Result<Vec<u8>> {
    let (stream, expected_size) = if data.starts_with(MAGIC) {
        if data.len() < 12 {
            bail!("truncated LZ77 header");
        }
        let expected = read_u32_le(data, 4) as usize;
        let comp_size = read_u32_le(data, 8) as usize;
        let stream_end = 12 + comp_size;
        if data.len() < stream_end {
            if strict {
                bail!(
                    "truncated LZ77 stream: have {} bytes, header comp_size={}",
                    data.len().saturating_sub(12),
                    comp_size
                );
            }
            (&data[12..], expected)
        } else {
            (&data[12..stream_end], expected)
        }
    } else {
        let expected = expected_size
            .ok_or_else(|| anyhow::anyhow!("expected_size or LZ77 header required"))?;
        (data, expected)
    };

    let mut out = Vec::with_capacity(expected_size);
    let mut window = [0_u8; WINDOW_SIZE];
    let mut window_pos = WINDOW_START;
    let mut pos = 0;

    while pos < stream.len() && out.len() < expected_size {
        let flags = stream[pos];
        pos += 1;

        for bit in 0..8 {
            if out.len() >= expected_size {
                break;
            }

            let is_literal = (flags & (1 << bit)) != 0;
            if is_literal {
                if pos >= stream.len() {
                    break;
                }
                let value = stream[pos];
                pos += 1;
                out.push(value);
                window[window_pos] = value;
                window_pos = (window_pos + 1) & 0xFFF;
            } else {
                if pos + 1 >= stream.len() {
                    break;
                }
                let b1 = stream[pos];
                let b2 = stream[pos + 1];
                pos += 2;

                let mut offset = b1 as usize | (((b2 & 0xF0) as usize) << 4);
                let length = ((b2 & 0x0F) as usize) + MIN_MATCH;

                for _ in 0..length {
                    if out.len() >= expected_size {
                        break;
                    }
                    let value = window[offset];
                    out.push(value);
                    window[window_pos] = value;
                    window_pos = (window_pos + 1) & 0xFFF;
                    offset = (offset + 1) & 0xFFF;
                }
            }
        }
    }

    if strict && out.len() != expected_size {
        bail!(
            "incomplete LZ77 decompression: output={} expected={}",
            out.len(),
            expected_size
        );
    }

    Ok(out)
}

pub fn compress_lz77(uncompressed: &[u8], all_literal: bool) -> Vec<u8> {
    let mut compressed = Vec::new();
    let mut window = [0_u8; WINDOW_SIZE];
    let mut window_pos = WINDOW_START;
    let mut src_pos = 0;
    let mut block_flags = 0_u8;
    let mut bit_count = 0;
    let mut block_tokens = Vec::new();

    while src_pos < uncompressed.len() {
        let max_len = MAX_MATCH.min(uncompressed.len() - src_pos);
        let (match_offset, mut match_len) = if !all_literal && max_len >= MIN_MATCH {
            find_best_match(uncompressed, src_pos, max_len, &window, window_pos)
        } else {
            (0, 0)
        };

        if match_len < MIN_MATCH {
            match_len = 1;
            block_flags |= 1 << bit_count;
            block_tokens.push(uncompressed[src_pos]);
        } else {
            let b1 = (match_offset & 0xFF) as u8;
            let b2 =
                (((match_offset >> 4) & 0xF0) as u8) | (((match_len - MIN_MATCH) & 0x0F) as u8);
            block_tokens.extend([b1, b2]);
        }

        for k in 0..match_len {
            window[window_pos] = uncompressed[src_pos + k];
            window_pos = (window_pos + 1) & 0xFFF;
        }

        src_pos += match_len;
        bit_count += 1;
        if bit_count == 8 {
            flush_block(
                &mut compressed,
                &mut block_flags,
                &mut bit_count,
                &mut block_tokens,
            );
        }
    }

    flush_block(
        &mut compressed,
        &mut block_flags,
        &mut bit_count,
        &mut block_tokens,
    );

    let mut out = Vec::with_capacity(12 + compressed.len());
    out.extend(MAGIC);
    out.extend((uncompressed.len() as u32).to_le_bytes());
    out.extend((compressed.len() as u32).to_le_bytes());
    out.extend(compressed);
    out
}

fn find_best_match(
    uncompressed: &[u8],
    src_pos: usize,
    max_len: usize,
    window: &[u8; WINDOW_SIZE],
    window_pos: usize,
) -> (usize, usize) {
    let search_start = src_pos.saturating_sub(WINDOW_SIZE);
    let search_end = src_pos.saturating_sub(MIN_MATCH);
    let search_range = (search_end.saturating_sub(search_start)).min(WINDOW_SIZE);

    let mut best_len = 0;
    let mut best_offset = 0;

    for dist in MIN_MATCH..=(search_range + 1).min(WINDOW_SIZE) {
        let w_idx = window_pos.wrapping_sub(dist) & 0xFFF;
        let search_max = max_len.min(dist);
        let mut curr_len = 0;
        while curr_len < search_max {
            let wpos = (w_idx + curr_len) & 0xFFF;
            if window[wpos] != uncompressed[src_pos + curr_len] {
                break;
            }
            curr_len += 1;
        }
        if curr_len > best_len {
            best_len = curr_len;
            best_offset = w_idx;
            if best_len == max_len {
                break;
            }
        }
    }

    (best_offset, best_len)
}

fn flush_block(
    compressed: &mut Vec<u8>,
    block_flags: &mut u8,
    bit_count: &mut usize,
    block_tokens: &mut Vec<u8>,
) {
    if *bit_count > 0 {
        compressed.push(*block_flags);
        compressed.append(block_tokens);
        *block_flags = 0;
        *bit_count = 0;
    }
}

fn read_u32_le(data: &[u8], offset: usize) -> u32 {
    u32::from_le_bytes(
        data[offset..offset + 4]
            .try_into()
            .expect("slice length checked"),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::datafat::{find_row, read_entries};

    #[test]
    fn literal_roundtrip() {
        let input = b"hello strawberry panic";
        let compressed = compress_lz77(input, true);
        let decompressed = decompress_lz77(&compressed, None, true).unwrap();
        assert_eq!(decompressed, input);
    }

    #[test]
    fn compressed_roundtrip() {
        let input = b"aaaaabbbbbccccccccccccccccccccccdddddaaaaabbbbb";
        let compressed = compress_lz77(input, false);
        let decompressed = decompress_lz77(&compressed, None, true).unwrap();
        assert_eq!(decompressed, input);
    }

    #[test]
    fn rejects_truncated_header() {
        assert!(decompress_lz77(b"LZ77", None, true).is_err());
    }

    #[test]
    fn decompresses_real_lz77_entry_when_available() {
        let path = std::path::Path::new("originales/Data.bin");
        if !path.exists() {
            return;
        }

        let rows = read_entries(path).unwrap();
        let data = std::fs::read(path).unwrap();
        let entry = rows
            .iter()
            .find(|entry| {
                if entry.offset == 0 || entry.size < 12 {
                    return false;
                }
                let start = entry.offset as usize;
                data.get(start..start + 4) == Some(&b"LZ77"[..])
            })
            .or_else(|| find_row(&rows, 7461));

        let Some(entry) = entry else { return };
        let start = entry.offset as usize;
        let end = start + entry.size as usize;
        if end > data.len() || data.get(start..start + 4) != Some(&b"LZ77"[..]) {
            return;
        }

        let decompressed = decompress_lz77(&data[start..end], None, true).unwrap();
        assert!(!decompressed.is_empty());
    }
}
